### PR TITLE
Added element forge

### PR DESCRIPTION
**Pull request by NCBPFluffyBear**

### CODE REVIEW

File: src/main/java/io/github/seggan/slimypowersuits/Setup.java [hunk 1]
```diff
@@ -19,13 +19,18 @@
...
+import org.bukkit.NamespacedKey;
 import org.bukkit.inventory.ItemStack;
 
 public class Setup {
 
     static void registerItems(SlimyPowerSuits plugin) {
+
+        new ElementForge(MiscItems.SUITS, MiscItems.ELEMENT_FORGE).register(plugin);
+
+        RecipeType ELEMENT_FORGE_RECIPE = new RecipeType(new NamespacedKey(SlimyPowerSuits.getInstance(), "element_forge_recipe"), MiscItems.ELEMENT_FORGE, "&cElement Forge", "&7Used to create powerful items");
```
Comment (Seggan) [inline src/main/java/io/github/seggan/slimypowersuits/Setup.java]: Eh, I don't know. Maybe something more useful like "Used to create new elements" or something?

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java [hunk 2]
```diff
@@ -22,6 +22,15 @@ private MiscItems() {}
...
 
+    public static final SlimefunItemStack ELEMENT_FORGE = new SlimefunItemStack(
+        "ELEMENT_FORGE",
+        Material.SMITHING_TABLE,
+        "&cElement Forge",
+        "",
+        "&7Used to create powerful items",
+        "&cMultiblock Structure"
+    );
+
```
Comment (Seggan) [inline src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: Same as aforementioned

File: src/main/java/io/github/seggan/slimypowersuits/Setup.java (hunk 1 above)
Comment (Seggan) [inline src/main/java/io/github/seggan/slimypowersuits/Setup.java]: Put this in element forge class and make a getter

**Review by Seggan — CHANGES_REQUESTED** [review]

**Comment by J3fftw1** [comment]
you should use the hash of the heads instead of the base64
https://www.base64decode.org/ decode them

**Comment by NCBPFluffyBear** [comment]
> you should use the hash of the heads instead of the base64
> https://www.base64decode.org/ decode them

I never used heads

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        Block block = null;
+
+        if (b.getRelative(1, 0, 0).getType() == Material.DISPENSER) {
+            block = b.getRelative(1, 0, 0);
+        } else if (b.getRelative(0, 0, 1).getType() == Material.DISPENSER) {
+            block = b.getRelative(0, 0, 1);
+        } else if (b.getRelative(-1, 0, 0).getType() == Material.DISPENSER) {
+            block = b.getRelative(-1, 0, 0);
+        } else if (b.getRelative(0, 0, -1).getType() == Material.DISPENSER) {
+            block = b.getRelative(0, 0, -1);
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: Material dispenser = getType() == Material.DISPENSER

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java [hunk 4]
```diff
@@ -13,115 +13,125 @@
...
     public static final SlimefunItemStack SUIT_GENERATOR = new SlimefunItemStack(
-            "SUIT_GENERATOR",
-            "eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvYTFkNWExZmY3Zjk3NmMxYzJlYmQ0ZWY5YTkwYWQ5MTQ2Nzk1YzFjNDRmZGFlNjI5NjQ5NDg0MzRhNzI1NyJ9fX0=",
-            "&6Power Suit Generator",
-            "",
-            "&7The central power",
-            "&7source of any power",
-            "&7suit."
+        "SUIT_GENERATOR",
+        "eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvYTFkNWExZmY3Zjk3NmMxYzJlYmQ0ZWY5YTkwYWQ5MTQ2Nzk1YzFjNDRmZGFlNjI5NjQ5NDg0MzRhNzI1NyJ9fX0=",
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: use the hash

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java [hunk 5]
```diff
@@ -13,115 +13,125 @@
...
     );
 
     public static final SlimefunItemStack SUIT_AI = new SlimefunItemStack(
-            "POWER_SUIT_AI",
-            "eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvZGM5MzY1NjQyYzZlZGRjZmVkZjViNWUxNGUyYmM3MTI1N2Q5ZTRhMzM2M2QxMjNjNmYzM2M1NWNhZmJmNmQifX19",
-            "&9Power Suit AI",
-            "",
-            "&7The object that contains",
-            "&7a power suit's AI."
+        "POWER_SUIT_AI",
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: use the hash

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java [hunk 6]
```diff
@@ -13,115 +13,125 @@
 
 public final class MiscItems {
 
-    public static final String MODULE_TEXTURE = "eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvZGZlYjM5ZDcxZWY4ZTZhNDI2NDY1OTMzOTNhNTc1M2NlMjZhMWJlZTI3YTBjYThhMzJjYjYzN2IxZmZhZSJ9fX0=";
+    public static final String MODULE_TEXTURE =
+        "eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvZGZlYjM5ZDcxZWY4ZTZhNDI2NDY1OTMzOTNhNTc1M2NlMjZhMWJlZTI3YTBjYThhMzJjYjYzN2IxZmZhZSJ9fX0=";
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: use the hash

File: src/main/java/io/github/seggan/slimypowersuits/Utils.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package io.github.seggan.slimypowersuits;
+
+import org.bukkit.Bukkit;
+import org.bukkit.scheduler.BukkitTask;
+
+public class Utils {
+    public static BukkitTask runSync(Runnable r, long delay) {
+        return SlimyPowerSuits.getInstance() != null && SlimyPowerSuits.getInstance().isEnabled() ? Bukkit.getScheduler().runTaskLater(SlimyPowerSuits.getInstance(), r, delay) : null;
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/Utils.java]: this is way to long formatting

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        return block;
+    }
+
+    private void movePiston(Block piston, Piston piss, boolean extended) {
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java:142]: dont call it piss...

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

**Comment by J3fftw1** [comment]
> 
> 
> > you should use the hash of the heads instead of the base64
> > https://www.base64decode.org/ decode them
> 
> I never used heads

you touched them so change it

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java (hunk 8 above)
Comment (TheBusyBiscuit) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java:142]: :LUL:

File: src/main/java/io/github/seggan/slimypowersuits/Utils.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package io.github.seggan.slimypowersuits;
+
+import org.bukkit.Bukkit;
+import org.bukkit.scheduler.BukkitTask;
+
+public class Utils {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/seggan/slimypowersuits/Utils.java]: Also this class should be final and have a private zero-args constructor.

**Comment by J3fftw1** [comment]
conflicts

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            block = b.getRelative(1, 0, 0);
+        } else if (b.getRelative(0, 0, 1).getType() == Material.DISPENSER) {
+            block = b.getRelative(0, 0, 1);
+        } else if (b.getRelative(-1, 0, 0).getType() == Material.DISPENSER) {
+            block = b.getRelative(-1, 0, 0);
+        } else if (b.getRelative(0, 0, -1).getType() == Material.DISPENSER) {
+            block = b.getRelative(0, 0, -1);
+        }
+
+        return block;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: No need for this Block variable, you could just immediately return the found value in each branch.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+        // Updating the Piston Head
+        if (extended) {
+            PistonHead head = (PistonHead) Material.PISTON_HEAD.createBlockData();
+            head.setFacing(BlockFace.DOWN);
+
+            piston.getRelative(BlockFace.DOWN).setBlockData(head, false);
+            piston.getWorld().playSound(piston.getLocation(), Sound.BLOCK_PISTON_CONTRACT, 0.5f, 0.5f);
+        } else {
+            piston.getRelative(BlockFace.DOWN).setType(Material.AIR);
```
Comment (J3fftw1) [inline src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java:154]: not sure how much better it would be but BlockFace.DOWN in a var?

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java:154]: I think this wouldnt matter as enum values are stored in one location

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java (hunk 6 above)
Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: i did that, i prefer base64

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java (hunk 5 above)
Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: i did that, i prefer base64

File: src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java (hunk 4 above)
Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/lists/MiscItems.java]: i did that, i prefer base64

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+ * @author NCBPFluffyBear
+ */
+public class ElementForge extends MultiBlockMachine {
+
+    public static final RecipeType ELEMENT_FORGE_RECIPE = new RecipeType(
+        new NamespacedKey(SlimyPowerSuits.getInstance(), "element_forge_recipe"),
+        MiscItems.ELEMENT_FORGE,
+        "&cElement Forge",
+        "&7Used to create new elements"
+    );
```
Comment (Seggan) [inline src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: Make a getter for this...

**Review by Seggan — CHANGES_REQUESTED** [review]

**Comment by Seggan** [comment]
Resolve ze conflicts.

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java (hunk 3 above)
Comment (NCBPFluffyBear) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: what

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: hmm idk; finish the other requests and ill think abt this one

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java (hunk 11 above)
Comment (NCBPFluffyBear) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java:154]: How much does this help performance wise? I feel like doing such a change would make it messier, not neater

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java:154]: > 
> 
> I think this wouldnt matter as enum values are stored in one location

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java (hunk 12 above)
Comment (NCBPFluffyBear) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: image
not sure how this is any better than the variable

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: no make the variable `private static final` and make a method `getRecipeType()` returning the variable

Comment (NCBPFluffyBear) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: getRecipeType() is a preexisting method, is this intentional

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: at least do 

>
> 
> make the variable `private static final` and make a method returning the variable

File: src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java (hunk 3 above)
Comment (J3fftw1) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: You do that functuon 3 times so make it a var

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: but what would u run `getType` on? give some example code

Comment (J3fftw1) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: Var this
Material dispenser = getType() == Material.DISPENSER
 else if (b.getRelative(0, 0, -1).dispenser

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: is that even possible in java

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: no...

Comment (Seggan) [reply src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java]: Hmm it seems that it's only possible with lambdas

**Review by Seggan — APPROVED** [review]
Sorry but I really need this now. I'll do the changes myself.

**Pull request merged**